build/scripts/build-utils.ps1

```diff
@@ -0,0 +1,151 @@
+# Collection of powershell build utility functions that we use across our scripts.
+
+set-strictmode -version 2.0
+$ErrorActionPreference="Stop"
+
+# Declare a number of useful variables for other scripts to use
+[string]$repoDir = [IO.Path]::GetFullPath($(join-path $PSScriptRoot "..\.."))
+[string]$binariesDir = join-path $repoDir "Binaries"
+[string]$scriptDir = $PSScriptRoot
+
+# The intent of this script is to locate and return the path to the MSBuild directory that
+# we should use for bulid operations.  The preference order for MSBuild to use is as 
+# follows
+#
+#   1. MSBuild from an active VS command prompt
+#   2. MSBuild from a machine wide VS install
+#   3. MSBuild from the xcopy toolset 
+#
+# This function will return two values: the kind of MSBuild chosen and the MSBuild directory.
+function locate-msbuild-core() {
+    param ([switch]$xcopy = $false) 
+
+    if ($xcopy) { 
+        write-output "xcopy"
+        write-output (locate-msbuild-xcopy)
+        return
+    }
+
+    # MSBuild from an active VS command prompt.  
+    if (${env:VSINSTALLDIR} -ne $null) {
+
+        # This line deliberately avoids using -ErrorAction.  Inside a VS command prompt
+        # an MSBuild command should always be available.
+        $p = (get-command msbuild).Source
+        write-output "vscmd"
+        write-output $p
+        return
+    }
+
+    # Look for a valid VS installation
+    try {
+        # It's important to execute locate-vs using a new powershell process.  This script will 
+        # use assemblies from our NuGet cache to locate VS.  This will cause them to be locked on
+        # disk for the duration of the powershell session.  Using a separate process makes sure
+        # they are not locked to this session which can be long lived.
+        $p = & powershell -noprofile -executionPolicy RemoteSigned -file (join-path $PSScriptRoot "locate-vs.ps1")
```

Comment:
There is now an 'official' tool (that is one I didn't write and that is in the Microsoft org, rather than a Roslyn specific repo): https://github.com/Microsoft/vswhere